editor/editor_properties.cpp

```diff
@@ -180,44 +180,150 @@ EditorPropertyMultilineText::EditorPropertyMultilineText() {
 
 ///////////////////// TEXT ENUM /////////////////////////
 
-void EditorPropertyTextEnum::_option_selected(int p_which) {
+void EditorPropertyTextEnum::_emit_changed_value(String p_string) {
 	if (string_name) {
-		emit_changed(get_edited_property(), StringName(options->get_item_text(p_which)));
+		emit_changed(get_edited_property(), StringName(p_string));
 	} else {
-		emit_changed(get_edited_property(), options->get_item_text(p_which));
+		emit_changed(get_edited_property(), p_string);
 	}
 }
 
+void EditorPropertyTextEnum::_option_selected(int p_which) {
+	_emit_changed_value(option_button->get_item_text(p_which));
+}
+
+void EditorPropertyTextEnum::_edit_custom_value() {
+	default_layout->hide();
+	edit_custom_layout->show();
+	custom_value_edit->grab_focus();
+}
+
+void EditorPropertyTextEnum::_custom_value_submitted(String p_value) {
+	edit_custom_layout->hide();
+	default_layout->show();
+
+	_emit_changed_value(p_value.strip_edges());
+}
+
+void EditorPropertyTextEnum::_custom_value_accepted() {
+	String new_value = custom_value_edit->get_text().strip_edges();
+	_custom_value_submitted(new_value);
+}
+
+void EditorPropertyTextEnum::_custom_value_cancelled() {
+	custom_value_edit->set_text(get_edited_object()->get(get_edited_property()));
+
+	edit_custom_layout->hide();
+	default_layout->show();
+}
+
 void EditorPropertyTextEnum::update_property() {
-	String which = get_edited_object()->get(get_edited_property());
-	for (int i = 0; i < options->get_item_count(); i++) {
-		String t = options->get_item_text(i);
-		if (t == which) {
-			options->select(i);
-			return;
+	String current_value = get_edited_object()->get(get_edited_property());
+	int default_option = options.find(current_value);
+
+	// The list can change in the loose mode.
+	if (loose_mode) {
+		custom_value_edit->set_text(current_value);
+		option_button->clear();
+
+		// Manually entered value.
+		if (default_option < 0 && !current_value.is_empty()) {
+			option_button->add_item(current_value, options.size() + 1001);
+			option_button->select(0);
+
+			option_button->add_separator();
 		}
+
+		// Add an explicit empty value for clearing the property.
+		option_button->add_item("", options.size() + 1000);
+
+		for (int i = 0; i < options.size(); i++) {
+			option_button->add_item(options[i], i);
+			if (options[i] == current_value) {
+				option_button->select(option_button->get_item_count() - 1);
+			}
+		}
+	} else {
+		option_button->select(default_option);
 	}
 }
 
-void EditorPropertyTextEnum::setup(const Vector<String> &p_options, bool p_string_name) {
+void EditorPropertyTextEnum::setup(const Vector<String> &p_options, bool p_string_name, bool p_loose_mode) {
+	string_name = p_string_name;
+	loose_mode = p_loose_mode;
+
+	options.clear();
+
+	if (loose_mode) {
+		// Add an explicit empty value for clearing the property in the loose mode.
+		option_button->add_item("", options.size() + 1000);
```

Comment:
What is the purpose of `options.size() + 1000`? Seems like the argument doesn't have any effect.

Reply:
Just so we don't conflict with the default options which for convenience have ids that match their original indices. It can just as well be `+ 1`, but a bigger number gives a nice offset, and I don't think it matters.